crates/paralegal-policy/src/flows_to.rs

```diff
@@ -13,31 +14,45 @@ pub struct CtrlFlowsTo {
     /// The indexes of [`DataSource`]s in the controller.
     pub sources: Arc<IndexedDomain<DataSource>>,
 
-    /// The indexes of [`DataSink`]s in the controller.
-    pub sinks: Arc<IndexedDomain<DataSink>>,
+    /// The indexes of [`CallSiteOrDataSink`]s in the controller.
+    pub sinks: Arc<IndexedDomain<CallSiteOrDataSink>>,
 
     /// The transitive closure of the [`Ctrl::data_flow`] relation.
+    /// If a source flows to a [`DataSink::Argument`], it also flows into its CallSite.
     ///
     /// See the [`IndexMatrix`] documentation for details on how to
     /// query this representation of the relation.
-    pub flows_to: IndexMatrix<DataSourceIndex, DataSink>,
+    pub data_flows_to: IndexMatrix<DataSourceIndex, CallSiteOrDataSink>,
+
+    /// The transitive closure of the [`Ctrl::data_flow`] and [`Ctrl::ctrl_flow`] relations.
+    /// If a source data-flows to a [`DataSink::Argument`], it flows into its CallSite.
+    /// If a source control-flows into a CallSite, it also flows into all of the [`DataSink::Argument`]s related to it.
+    pub flows_to: IndexMatrix<DataSourceIndex, CallSiteOrDataSink>,
 }
 
 impl CtrlFlowsTo {
     /// Constructs the transitive closure from a [`Ctrl`].
     pub fn build(ctrl: &Ctrl) -> Self {
         // Collect all sources and sinks into indexed domains.
-        let sources = Arc::new(IndexedDomain::from_iter(ctrl.data_flow.0.keys().cloned()));
-        let sinks = Arc::new(IndexedDomain::from_iter(
-            ctrl.data_flow.0.values().flatten().dedup().cloned(),
+        let sources = Arc::new(IndexedDomain::from_iter(
+            ctrl.all_sources().iter().map(|&s| s.clone()),
         ));
+        let sinks = Arc::new(IndexedDomain::from_iter(ctrl.all_call_sites_or_sinks()));
 
         // Connect each function-argument sink to its corresponding function sources.
         // This lets us compute the transitive closure by following through the `sink_to_source` map.
         let mut sink_to_source = IndexMatrix::new(&sources);
         for (sink_idx, sink) in sinks.as_vec().iter_enumerated() {
             for (src_idx, src) in sources.as_vec().iter_enumerated() {
-                if let (DataSource::FunctionCall(f1), DataSink::Argument { function: f2, .. }) =
+                if let (
```

Comment:
Since both of the `if let`'s match on `(src, sink)` a `match` block would be easier to read here I think.

Reply:
Done